llama.cpp

```diff
@@ -13615,21 +13624,26 @@ void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * c
         }
     }
 
+    // Store decoded codepoints when they are not cached (happens when there's a partial utf8 string prefix).
     std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>> candidates_decoded;
-    candidates_decoded.reserve(candidates->size);
+    if (grammar->partial_utf8.n_remain > 0) {
+        candidates_decoded.reserve(candidates->size);
```

Comment:
Almost feels like -- in the case of no partial_utf8 -- we could almost just copy the entire `ctx->token_codepoints_without_partial_utf8_prefix` array into `candidates_decoded` and call it a day -- is there a reason to rebuild this by calling `push_back` on every element individually?

Probably not a huge time-save, but just thinking out loud as I review this code again.

Reply:
Yes! And that's what is being done a few lines below in the `if (grammar->partial_utf8.n_remain == 0){` branch :-) In case of non partial_utf8 it skips the candidates_decoded step and just references the cached entries in the element pushed to candidates_grammar (no copies needed). Tried to comment this above but maybe it deserves a name change too, lemme know.

Reply:
Oh whoops, I mixed up candidates_decoded with candidates_grammar.

I think what I was imagining was something like:

```
    std::vector<llama_grammar_candidate>                              candidates_grammar;
    // Store decoded codepoints when they are not cached (happens when there's a partial utf8 string prefix).
    std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>> candidates_decoded;
    if (grammar->partial_utf8.n_remain > 0) {
        candidates_decoded.reserve(candidates->size);
        candidates_grammar.reserve(candidates->size);
    } else {
        // If there are no partial utf8 strings, we can use the cached codepoints directly.
        // Copy directly from ctx->token_codepoints_without_partial_utf8_prefix to candidates_grammar and ignore candidates_decoded.
        candidates_grammar.assign(ctx->token_codepoints_without_partial_utf8_prefix.begin(), ctx->token_codepoints_without_partial_utf8_prefix.end());
    }
```
And then I was imagining that one could skip the for-loop entirely (or at least greatly shorten it).

However, after reading through things again, that doesn't work for a few reasons -- not every token needs to be added to `candidates_grammar`.

Also, `token_codepoints_without_partial_utf8` is of type `std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>>`, and `candidates_grammar` is of type `std::vector<llama_grammar_candidate>`, so the data types don't line up.

Though speaking of which, I wonder if the data type of `token_codepoints_without_partial_utf8` should be changed to match `candidates_grammar`? It would simplify the line a little lower to be:

```
        } else if (grammar->partial_utf8.n_remain == 0){
            const auto & decoded = ctx->token_codepoints_without_partial_utf8_prefix.at(id);
            candidates_grammar.push_back(decoded);
```

?

Reply:
> Though speaking of which, I wonder if the data type of token_codepoints_without_partial_utf8 should be changed to match candidates_grammar?

Answering my own question, it would simplify that particular line, but it would complicate the changes in this section:

```
    // Note terminating 0 in decoded string
    const auto   decoded     = grammar->partial_utf8.n_remain == 0
        ? ctx->token_codepoints_without_partial_utf8_prefix[token]
        : decode_utf8(piece, grammar->partial_utf8);
```

I think the way you have it here is fine -- don't worry about changing anything!